File: src/test/java/com/koddy/server/member/application/usecase/CompleteProfileUseCaseTest.java
package com.koddy.server.member.application.usecase;

import com.koddy.server.common.UnitTest;
import com.koddy.server.common.fixture.MentoringPeriodFixture;
import com.koddy.server.common.fixture.TimelineFixture;
import com.koddy.server.member.application.usecase.command.CompleteMenteeProfileCommand;
import com.koddy.server.member.application.usecase.command.CompleteMentorProfileCommand;
import com.koddy.server.member.domain.model.mentee.Mentee;
import com.koddy.server.member.domain.model.mentor.Mentor;
import com.koddy.server.member.domain.repository.MenteeRepository;
import com.koddy.server.member.domain.repository.MentorRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_1;
import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@DisplayName("Member -> CompleteProfileUseCase 테스트")
class CompleteProfileUseCaseTest extends UnitTest {
    private final MentorRepository mentorRepository = mock(MentorRepository.class);
    private final MenteeRepository menteeRepository = mock(MenteeRepository.class);
    private final CompleteProfileUseCase sut = new CompleteProfileUseCase(mentorRepository, menteeRepository);

    @Test
    @DisplayName("멘토 프로필을 완성한다")
    void completeMentor() {
        // given
        final Mentor mentor = new Mentor(
                MENTOR_1.getEmail(),
                MENTOR_1.getName(),
                MENTOR_1.getProfileImageUrl(),
                MENTOR_1.getLanguages(),
                MENTOR_1.getUniversityProfile()
        ).apply(1L);
        assertAll(
                () -> assertThat(mentor.getIntroduction()).isNull(),
                () -> assertThat(mentor.getMentoringPeriod()).isNull(),
                () -> assertThat(mentor.getSchedules()).isEmpty(),
                () -> assertThat(mentor.isProfileComplete()).isFalse()
        );

        final CompleteMentorProfileCommand command = new CompleteMentorProfileCommand(
                mentor.getId(),
                MENTOR_1.getIntroduction(),
                MentoringPeriodFixture.FROM_03_01_TO_05_01.toDomain(),
                TimelineFixture.월_수_금()
        );
        given(mentorRepository.getById(command.mentorId())).willReturn(mentor);

        // when
        sut.completeMentor(command);

        // then
        assertAll(
                () -> verify(mentorRepository, times(1)).getById(command.mentorId()),
                () -> verify(menteeRepository, times(0)).getById(anyLong()),
                () -> assertThat(mentor.getIntroduction()).isNotNull(),
                () -> assertThat(mentor.getMentoringPeriod()).isEqualTo(command.mentoringPeriod()),
                () -> assertThat(mentor.getSchedules()).hasSize(TimelineFixture.월_수_금().size()),
                () -> assertThat(mentor.isProfileComplete()).isTrue()
        );
    }

    @Test
    @DisplayName("멘티 프로필을 완성한다")
    void completeMentee() {
        // given
        final Mentee mentee = new Mentee(
                MENTEE_1.getEmail(),
                MENTEE_1.getName(),
                MENTEE_1.getProfileImageUrl(),
                MENTEE_1.getNationality(),
                MENTEE_1.getLanguages(),
                MENTEE_1.getInterest()
        ).apply(1L);
        assertAll(
                () -> assertThat(mentee.getIntroduction()).isNull(),
                () -> assertThat(mentee.isProfileComplete()).isFalse()
        );

        final CompleteMenteeProfileCommand command = new CompleteMenteeProfileCommand(mentee.getId(), MENTEE_1.getIntroduction());
        given(menteeRepository.getById(command.menteeId())).willReturn(mentee);

        // when
        sut.completeMentee(command);

        // then
        assertAll(
                () -> verify(mentorRepository, times(0)).getById(anyLong()),
                () -> verify(menteeRepository, times(1)).getById(command.menteeId()),
                () -> assertThat(mentee.getIntroduction()).isNotNull(),
                () -> assertThat(mentee.isProfileComplete()).isTrue()
        );
    }
}

File: src/test/java/com/koddy/server/member/application/usecase/CompleteProfileUseCaseTest.kt
package com.koddy.server.member.application.usecase

import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.fixture.MenteeFixtureStore.menteeFixture
import com.koddy.server.common.fixture.MentorFixtureStore.mentorFixture
import com.koddy.server.member.application.usecase.command.CompleteMenteeProfileCommand
import com.koddy.server.member.application.usecase.command.CompleteMentorProfileCommand
import com.koddy.server.member.domain.model.mentee.Mentee
import com.koddy.server.member.domain.model.mentor.Mentor
import com.koddy.server.member.domain.service.MemberReader
import io.kotest.assertions.assertSoftly
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify

@UnitTestKt
@DisplayName("Member -> CompleteProfileUseCase 테스트")
internal class CompleteProfileUseCaseTest : DescribeSpec({
    val memberReader = mockk<MemberReader>()
    val sut = CompleteProfileUseCase(memberReader)

    val mentorFixture = mentorFixture(id = 1L)
    val menteeFixture = menteeFixture(id = 2L)

    describe("CompleteProfileUseCase's completeMentor") {
        context("멘토의 미완성 프로필에 대해서") {
            val mentor = Mentor(
                id = mentorFixture.id,
                platform = mentorFixture.platform,
                name = mentorFixture.name,
                languages = mentorFixture.languages,
                universityProfile = mentorFixture.universityProfile,
            )
            assertSoftly(mentor) {
                introduction shouldBe null
                profileImageUrl shouldBe null
                mentoringPeriod shouldBe null
                schedules shouldBe emptyList()
                profileComplete shouldBe false
            }

            val command = CompleteMentorProfileCommand(
                mentorId = mentorFixture.id,
                introduction = mentorFixture.introduction,
                profileImageUrl = mentorFixture.profileImageUrl,
                mentoringPeriod = mentorFixture.mentoringPeriod,
                timelines = mentorFixture.timelines,
            )
            every { memberReader.getMentor(command.mentorId) } returns mentor

            it("추가 정보를 기입한다 [자기소개, 프로필 이미지 URL, 멘토링 기간, 스케줄 정보]") {
                sut.completeMentor(command)

                verify(exactly = 1) { memberReader.getMentor(command.mentorId) }
                assertSoftly(mentor) {
                    introduction shouldBe command.introduction
                    profileImageUrl shouldBe command.profileImageUrl
                    mentoringPeriod!!.startDate shouldBe command.mentoringPeriod!!.startDate
                    mentoringPeriod!!.endDate shouldBe command.mentoringPeriod!!.endDate
                    mentoringPeriod!!.timeUnit shouldBe command.mentoringPeriod!!.timeUnit
                    schedules shouldHaveSize command.timelines.size
                    profileComplete shouldBe true
                }
            }
        }
    }

    describe("CompleteProfileUseCase's completeMentee") {
        context("멘티의 미완성 프로필에 대해서") {
            val mentee = Mentee(
                id = menteeFixture.id,
                platform = menteeFixture.platform,
                name = menteeFixture.name,
                nationality = menteeFixture.nationality,
                languages = menteeFixture.languages,
                interest = menteeFixture.interest,
            )
            assertSoftly(mentee) {
                introduction shouldBe null
                profileImageUrl shouldBe null
                profileComplete shouldBe false
            }

            val command = CompleteMenteeProfileCommand(
                menteeId = menteeFixture.id,
                introduction = menteeFixture.introduction,
                profileImageUrl = menteeFixture.profileImageUrl,
            )
            every { memberReader.getMentee(command.menteeId) } returns mentee

            it("추가 정보를 기입한다 [자기소개, 프로필 이미지 URL]") {
                sut.completeMentee(command)

                verify(exactly = 1) { memberReader.getMentee(command.menteeId) }
                assertSoftly(mentee) {
                    introduction shouldBe command.introduction
                    profileImageUrl shouldBe command.profileImageUrl
                    profileComplete shouldBe true
                }
            }
        }
    }
})
